### PR TITLE
Testcase OCS-260: Create Pod with Multiple PVCs attached

### PR DESCRIPTION
**Pull request by ramkiperiy**
new file:   ocs_ci/templates/app-pods/cirros.yaml
new file:   tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py
  -- Testcase OCS-260: Create pod with multiple PVC attached
  -- templates/app-pods: adding cirros.yaml, new app pod yaml file
  -- Added DeploymentConfig and Cirros.yaml in constants.py file
Signed-off(rperiyas@redhat.com)

### CODE REVIEW

**Comment by ramkiperiy** [comment]
Test case Run Passed
17:29:28 - MainThread - tests.manage.test_create_pod_attaching_multiple_PVC_start_IO - INFO - Running IO's for 212 seconds
PASSED                                                                                                                                                                [100%]
----------------------------------------------------------------------------- live log teardown -----------------------------------------------------------------------------
17:33:00 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete DeploymentConfig cirrospod
17:33:01 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig get PersistentVolumeClaim  -o yaml
17:33:03 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete PersistentVolumeClaim cirrospvc0
17:33:35 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete PersistentVolumeClaim cirrospvc1
17:33:37 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete PersistentVolumeClaim cirrospvc2
17:33:39 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete PersistentVolumeClaim cirrospvc3
17:33:41 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete PersistentVolumeClaim cirrospvc4
17:33:43 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete StorageClass storageclass-test-0217270237
17:33:44 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete Secret secret-test-0217265005
17:33:46 - MainThread - ocs_ci.utility.utils - INFO - Executing command: oc -n openshift-storage --kubeconfig /home/rperiyas/2ndJuly/auth/kubeconfig delete CephBlockPool cirrostest
======================================================================== 1 passed in 655.84 seconds =========================================================================

File: tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    ocs_obj = OCS(**cirros_yaml_data)
+    assert ocs_obj.create()
+    try:
+        # Check for pod is up and running
+        pod = ocp.OCP(
+            kind=constants.POD, namespace=config.ENV_DATA['cluster_namespace']
+        )
+        pod.wait_for_resource(
+            condition='Running', selector='deploymentconfig=cirrospod',
+            timeout=180, sleep=20
```
Comment (RazTamir) [inline tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py]: create constants for 180 and 20

File: tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    cirros_yaml_data['metadata']['namespace'] = config.ENV_DATA['cluster_namespace']
+    temp_list1 = cirros_yaml_data['spec']['template']['spec']['containers'][0]['args']
+    temp_list1.pop(len(temp_list1) - 1)
+    temp_list2 = cirros_yaml_data['spec']['template']['spec']['containers'][0][
+        'livenessProbe'
+    ]['exec']['command']
+    temp_list2.pop(len(temp_list2) - 1)
+    # for loop which creates the values, which will be added in cirros.yaml
+    for pv in pvc_list:
+        temp_dict1.append(
```
Comment (RazTamir) [inline tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py]: please create a function to 'attach_pvc_to_pod()' under pod.py or even better, 'attch_pvc()' directly under Pod object in pod.py

File: tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    pvc_list = []
+
+    def test_create_cirros_pod(self):
+        """
+        Test Creating Cirros POD with multiple PVC attached to it.
+        """
+        self.pvc_list = create_pvc(pvc_name_prefix="cirrospvc", pvc_count=self.pvc_count)
+        self.cirros_name.append(create_pod(self.pvc_list))
+        log.info(f"Running IO's for {self.io_duration_sec} seconds")
+        time.sleep(self.io_duration_sec)
```
Comment (RazTamir) [inline tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py]: is this a TODO to run IOs?

File: ocs_ci/templates/app-pods/cirros.yaml [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        image: 'cirros'
+        imagePullPolicy: 'IfNotPresent'
+        volumeMounts:
+        - mountPath: '/mnt'
+          name: 'cirros-vol'
+        command: ['sh']
+        args:
+        - '-ec'
+        - 'while true; do
+               (mount | grep /mnt) && (head -c 1000 < /dev/urandom > /mnt/random-data.log) || exit 1;
```
Comment (vasukulkarni) [inline ocs_ci/templates/app-pods/cirros.yaml]: you are doing the same in code, why is this needed again?

File: tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        interface_name=RBD_POOL.name,
+        secret_name=RBD_POOL_SECRET_OBJ.name
+    )
+
+
+def teardown(self):
+    """
+    Tearing down the environment
+    """
+    dc = ocp.OCP(
```
Comment (ebenahar) [inline tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py]: Please consider using ocs_ci.ocs.resources.pod.Pod or ocs_ci.ocs.resources.ocs.OCS instead

File: tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        kind=constants.DEPLOYMENTCONFIG, namespace=config.ENV_DATA['cluster_namespace']
+    )
+    dc.delete(resource_name=self.cirros_name[0], wait=True)
+    pvc.delete_all_pvcs()
+    RBD_STORAGE_CLASS_OBJ.delete()
+    RBD_POOL_SECRET_OBJ.delete()
+    RBD_POOL.delete()
+
+
+def create_pvc(pvc_name_prefix, pvc_count):
```
Comment (ebenahar) [inline tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py]: we have this:
ocs_ci.ocs.resources.pvc.create_multiple_pvc

File: tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    """
+    pvc_list = []
+    for count in range(pvc_count):
+        name = pvc_name_prefix + str(count)
+        pvc_obj = helpers.create_pvc(sc_name=RBD_STORAGE_CLASS_OBJ.name, pvc_name=name)
+        pvc_list.append(pvc_obj.name)
+    return pvc_list
+
+
+def create_pod(pvc_list):
```
Comment (ebenahar) [inline tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py]: I think you can break down the implementation here to pod creation (which you can take from ocs_ci.ocs.resources.pod.Pod) and as Raz suggested below, attach pvc to pod one

File: ocs_ci/templates/app-pods/cirros.yaml (hunk 4 above)
Comment (ramkiperiy) [reply ocs_ci/templates/app-pods/cirros.yaml]: @vasukulkarni This yaml file attaches the PVC and run dd IO continuously till the POD is deleted. Also this pod has liveliness probe enabled (line 30), so deleting this pod is actually respin of POD and after respin again IO will continue.

File: tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    del cirros_yaml_data['spec']['template']['spec']['containers'][0][
+        'livenessProbe'
+    ]['exec']['command']
+    # Updating the yaml with pv_list arg values, which will be mounted
+    # in the cirros pod.
+    cirros_yaml_data['spec']['template']['spec']['volumes'] = temp_dict1
+    cirros_yaml_data['spec']['template']['spec']['containers'][0][
+        'volumeMounts'
+    ] = temp_dict2
+    temp_list1.append('while true; do ' + temp_str + ' sleep 20 ; done')
```
[the remark this answers is not in the document]
Comment (ramkiperiy) [reply tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py]: @vasukulkarni, Am trying to add more PVC to the attached yaml based on test requirement, the sample yaml file has only one PVC, so part of the POD creation am editing the yaml file to add more PVC, and in each PVC IO will be started. 
In this piece of code am updating yaml line(line number 27) to add more PVC, so that IO starts and it will continue till the POD is deleted.

File: ocs_ci/templates/app-pods/cirros.yaml (hunk 4 above)
Comment (vasukulkarni) [reply ocs_ci/templates/app-pods/cirros.yaml]: @ramkiperiy thanks for explanation.

File: tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (vasukulkarni) [reply tests/manage/test_create_pod_attaching_multiple_PVC_start_IO.py]: got it thanks!

**Comment by vasukulkarni** [comment]
overall LGTM

**Comment by vasukulkarni** [comment]
@ramkiperiy some conflicts,  If you can address or resolve them with comments, we can merge this fast.

**Comment by ramkiperiy** [comment]
Planning to remove the dd IO part in cirros yaml, will try to run fio. The challenge is how to run fio on single cirros pod multiple mounts, and will update the PR with latest changes.

**Comment by RazTamir** [comment]
Hi Ram,
Google might help on that. 
In case no luck, maybe we can run multiple ios, on each mount?

**Comment by vasukulkarni** [comment]
@ramkiperiy let me know if you need help on this one, we can finish it up and merge this!

**Comment by RazTamir** [comment]
@ramkiperiy do we still need this?

**Comment by vasukulkarni** [comment]
@ramkiperiy ping?

**Comment by ramkiperiy** [comment]
Closing this PR. After this PR there are lot of changes happend in the framework. Test case requires lot of changes.
